[RFC][Routing] Added the Route attribute

derrabus opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | symfony/symfony-docs#14230

I was wondering if we can make use of [attributes](https://wiki.php.net/rfc/attributes_v2) as a replacement for Doctrine Annotations for configuring routes.

I have modified the existing `AnnotationClassLoader` so that it merges attributes and annotations. This way, an application could transparently switch from annotations to attributes. Moreover, the `AnnotationClassLoader` does not require an annotation reader anymore.

Since the chosen syntax `#[…]` is treated as a comment by php 7, I was able to use the existing annotation class, which makes the migration even easier.

#### Example with Doctrine Annotations
```php
use Symfony\Component\Routing\Annotation\Route;

class ActionPathController
{
    /**
     * @Route("/path", name: "action")
     */
    public function action()
    {
    }
}
```

#### Same example with the proposed Attribute
```php
use Symfony\Component\Routing\Annotation\Route;

class ActionPathController
{
    #[Route('/path', name: 'action')]
    public function action()
    {
    }
}
```

ro0NL reviewed (commented):
i like it :)

~~should we offer hybrid support by default? Thus vendor controllers could specify both attributes as well as annotations. Not sure this makes sense in practice though, otherwise i'd expect an exception if attributes and annotations are mixed&matched.~~
edit: nevermind, this works as expected already.

we could use sibling/nested attributes for rich options perhaps. Not sure it works out.

derrabus commented on src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php:
```diff
@@ -331,4 +340,32 @@ protected function createRoute(string $path, array $defaults, array $requirement
     abstract protected function configureRoute(Route $route, \ReflectionClass $class, \ReflectionMethod $method, $annot);
+
+    /**
+     * @param \ReflectionClass|\ReflectionMethod $reflection
```
One fine day, yes. But for now, this class needs to remain compatible with php 7.2.

ro0NL replied:
right -.-

derrabus commented:
Right now, you can mix&match attributes and annotations. It would have the same effect as if you specified multiple annotations. This was the easiest way to implement it, but this behavior is certainly debatable.

ro0NL commented:
i think it's fair to assume both definitions remain equivalent. Then it works on all SF versions, loading either one or another.

> I was wondering if we can make use of attributes as a replacement for Doctrine Annotations

raising the bigger question, is there any reason not to? Should we deprecate annotation support once we support php:^8?

Taluu commented on src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php:
```diff
@@ -331,4 +340,32 @@ protected function createRoute(string $path, array $defaults, array $requirement
+    {
+        if (method_exists($reflection, 'getAttributes')) {
+            foreach ($reflection->getAttributes($this->routeAttributeClass) as $attribute) {
+                yield $attribute->newInstance();
```
isn't it dangerous to use a `newInstance` ? or are the args taken from the attribute itself (didn't look that part)

Taluu replied:
ehh nevermind, saw the tests

derrabus replied:
Coming from Doctrine annotations, this looks odd, I admit. But the reflection API does not instantiate the actual attribute unless you tell it to.

javiereguiluz commented:
Please note that the attribute syntax has just changed from `<< ... >>` to `@@...`  See https://wiki.php.net/rfc/shorter_attribute_syntax

derrabus commented:
@javiereguiluz This has not been merged yet, has it?

javiereguiluz commented:
Not merged yet ... but approved, so it will be merged "soon" (I was notified about this by this nice @beberlei service: https://php-rfc-watch.beberlei.de/)

derrabus commented:
All right. If I changed the PR now, Travis won't be able to run my tests anymore. I don't expect the reflection API to be affected by the change, so all I will need to do is adjust the test fixtures and the attribute class itself to the new syntax. I'll do this as soon as Travis' nightly build supports it.

nicolas-grekas commented on src/Symfony/Component/Routing/Tests/Loader/AnnotationClassLoaderTest.php:
```diff
@@ -107,6 +117,18 @@ public function testInvokableControllerLoader()
+        $this->assertCount(1, $routes);
+        $this->assertEquals('/here', $routes->get('lol')->getPath());
+        $this->assertEquals(['GET', 'POST'], $routes->get('lol')->getMethods());
+        $this->assertEquals(['https'], $routes->get('lol')->getSchemes());
```
assertSame everywhere :)

derrabus replied:
This should be fixed on the original test case then. This is merely C&P for now.

My plan is duplicate the fixtures and rewrite the annotations to attributes, so we can run the same test against both fixture sets.

derrabus commented:
Pending PR for the syntax change: php/php-src#5796

derrabus commented:
I've pushed an update to this PR the other day.

* The PR is now incompatible with 8.0.0alpha1 and requires the changes of php/php-src#5751 to run.
    * `PhpAttribute` has been renamed to `Attribute`
    * The allowed targets of the `Route` attribute have been set to class and method (just like the annotation).
    * Now, we need to explicitly allow the attribute to be applied multiple times to the same structure, so I did just that.
* I've duplicated all test fixtures of `AnnotationClassLoaderTest` and rewrote the annotations to attributes. So we're running the same test suite for annotations and attributes now.

Some tests mocked Doctrine's annotation reader. I rewrote those test to use an actual annotation reader instead. It might be a good idea to backport that change, in order to ease the maintenance.

derrabus commented:
I might need some help with getting Travis green: https://travis-ci.org/github/symfony/symfony/jobs/705695329

On the php 7.4 run, it looks like some piece of code is trying to parse my new attribute (which only works on php 8). Anything I can do to prevent that?

nicolas-grekas commented:
patch `.github/patch-types.php`

derrabus commented:
> patch `.github/patch-types.php`?

That did not help, apparently. :-(

moliata commented:
@javiereguiluz Just to chime in, `@@` is still in "unknown status" given that it raises ambiguity in the PHP parser. Thus, whether the new syntax will be used depends on the [namespaced names as tokens RFC](https://wiki.php.net/rfc/namespaced_names_as_token).

derrabus commented:
PHP attributes are basically a moving target at the moment. Just look at the changes between the alpha1 and alpha2 releases. And because I need a working testsuite, I won't anticipate any syntax changes until they actually hit the php 8 nightly build on travis.

Nevertheless, I think working with attributes at this early stage can be very helpful:

* We can find out if attributes are helpful and how we can make use of them in Symfony.
* We can give early feedback to the php dev team, which in turn might help them to fine-tune the feature before the final release.

Since Symfony 5.2 will stabilize together with php 8, I'm marking this PR as ready now. The current implementation of the PR works find with the alpha2 of php 8. If the PR gets merged, I'll monitor the php 8 build on Travis and adjust the implementation if necessary.

derrabus commented:
I've pushed a small update. The `Route` attribute class can be used as a Doctrine-style annotation now. This should make the transition from annotations to attributes even smoother, because we can now convert the annotations one by one without importing both classes: Just use the attribute class. And one fine day, when Symfony drops php 7 support (😱) we could even merge the two classes.

I originally wanted to use the same class for annotations and attributes. The only thing that prevented me from doing so is that I have to configure the attribute class itself by attaching an attribute to it. And I cannot do this without breaking php 7 compatibility. This is why I've created the attribute as a subclass of the annotation class.

derrabus commented:
Hello again. I've updated the PR in order to make it compatible with php 8.0.0-beta1. The attribute syntax has changed from `<<Route>>` to `@@Route` as @javiereguiluz has already anticipated.

Moreover, php now supports named arguments which is a perfect match for attributes! I've updated the constructor so we don't have to work with those bulky arrays anymore. The syntax of the attribute is now almost the same as for doctrine annotations.

javiereguiluz commented:
@derrabus it looks lie `@@...` syntax could be dropped before the final release of PHP 8. The syntax is still being debated and the new candidates are `#[...]` and `@[...]` See https://externals.io/message/111312

derrabus commented:
@javiereguiluz Yes, I'm aware. I'll keep the PR updated.

derrabus commented:
I've cherry-picked #37771 into this PR in order to have the new tests executed on php 8. Please note that Doctrine annotations are currently broken due to doctrine/annotations#339.

derrabus committed:
[Routing] Added the Route attribute.

derrabus commented:
Another syntax change has hit the php master branch: https://github.com/php/php-src/commit/8b37c1e993f17345312a76a239b77ae137d3f8e8

Let's hope it's the final one. 🤞 

Because with the new `#[…]` syntax attributes can be written in a way so that they are treated as a comment by php 7, there is no need for a dedicated attribute class anymore. This is why I've merged my attribute class with the annotation class.

Travis' nightly php 8 build is a bit behind and still expects the `@@` syntax, so tests are failing currently. I'll trigger a new build in a couple of days when Travis has caught up.

Fabbot is red because it tries to "fix" the attribute from `#` to `//`. This is probably something we should report to the PHP CS Fixer project, I guess.

Taluu reviewed (commented):
For clarity's sake, I wouldn't mix the attribute version and the doctrine annot version, which would allow us to remove the data parameter and use proper typehinting, and be less a burden to maintain.

But this should be good enough. :D

Neirda24 commented on src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php:
```diff
@@ -61,14 +77,14 @@ abstract class AnnotationClassLoader implements LoaderInterface
     protected $defaultRouteIndex = 0;
 
-    public function __construct(Reader $reader)
+    public function __construct(Reader $reader = null)
```
Shouldn't it be `?Reader $reader = null` ?

fabpot commented on src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php:
```diff
@@ -257,7 +269,15 @@ protected function getGlobals(\ReflectionClass $class)
 
-        if ($annot = $this->reader->getClassAnnotation($class, $this->routeAnnotationClass)) {
+        $annot = null;
+        if (method_exists($class, 'getAttributes') && ($attribute = $class->getAttributes($this->routeAnnotationClass)[0] ?? null)) {
```
\PHP_VERSION_ID >= 80000

fabpot commented on src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php:
```diff
@@ -330,5 +350,33 @@ protected function createRoute(string $path, array $defaults, array $requirement
+     */
+    private function getAnnotations(object $reflection): iterable
+    {
+        if (method_exists($reflection, 'getAttributes')) {
```
\PHP_VERSION_ID >= 80000

fabpot commented:
Thank you @derrabus.

TomasVotruba commented:
We've just added `@Route` attribute to @rectorphp:
https://github.com/rectorphp/rector/pull/4350

Named parameters makes this very intuitive to migrate :+1: 


Note, there is no `Symfony\Component\Routing\Attribute\Route` class as in PR description.
The class is the same as before, `..\Annotation\Route`.

derrabus commented:
@TomasVotruba Thank you, I have updated the PR description. The first two iterations of the attribute syntax required me to introduce a dedicated attribute class, so this might be a leftover from that time.

TomasVotruba commented:
@derrabus Thanks :+1: 
I understand that, I rarely update the PR descriptions after design changes in them.
